### PR TITLE
[Feature] Add MobileOne Backbone

### PR DESCRIPTION
**Pull request by Ezra-Yu**
## Motivation

Add MobileOne backbone, 
[Paper](https://arxiv.org/abs/2206.04040)
[Github](https://github.com/apple/ml-mobileone)

## Note

*Because the [official repo.](https://github.com/apple/ml-mobileone) does not give a strategy for training and testing, the test data pipline of [RepVGG](https://github.com/open-mmlab/mmclassification/tree/master/configs/repvgg) is used here, and the result is about 0.1 lower than the official one. Refer to [this issue](https://github.com/apple/ml-mobileone/issues/2).*

## Use cases (Optional)
```
    Example:
        >>> from mmcls.models import MobileOne
        >>> import torch
        >>> model = MobileOne("s0", out_indices=(0, 1, 2, 3))
        >>> model.eval()
        >>> x = torch.rand(1, 3, 224, 224)
        >>> outputs = model(x)
        >>> for out in outputs:
        ...     print(tuple(out.shape))
        (1, 48, 56, 56)
        (1, 128, 28, 28)
        (1, 256, 14, 14)
        (1, 1024, 7, 7)
```
## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmcls/models/backbones/mobileone.py [hunk 1]
```diff
@@ -0,0 +1,514 @@
...
+    """MobileOne block for MobileOne backbone.
+
+    Args:
+        in_channels (int): The input channels of the block.
+        out_channels (int): The output channels of the block.
+        kernel_size (int): The kernel size of the convs in the block. If the
+            kernel size is large than 1, there will be a ``branch_scale`` in
+             the block.
+        num_convs (int): Number of conv branches in the block.
+        stride (int): Stride of convolution layers. Default: 1.
```
Comment (mzr1996) [inline mmcls/models/backbones/mobileone.py:27]: Please check all default values in the docstring, and use `defaults to xxx`.

File: mmcls/models/backbones/mobileone.py [hunk 2]
```diff
@@ -0,0 +1,514 @@
...
+        padding (int): Padding of the convolution layers. Default: 1.
+        dilation (int): Dilation of the convolution layers. Default: 1.
+        groups (int): Groups of the convolution layers. Default: 1.
+        se_cfg (None or dict): The configuration of the se module.
+            Default: None.
+        norm_cfg (dict): dictionary to construct and config norm layer.
+            Default: dict(type='BN', requires_grad=True).
```
Comment (mzr1996) [inline mmcls/models/backbones/mobileone.py:34]:
```suggestion
            Defaults to ``dict(type='BN', requires_grad=True)``.
```

File: mmcls/models/backbones/mobileone.py [hunk 3]
```diff
@@ -0,0 +1,514 @@
...
+
+    Args:
+        arch (str | dict): MobileOne architecture. If use string, choose
+            from 's0', 's1', 's2', 's3' and 's4'. If use dict, it should
+            have below keys:
+
+            - num_blocks (Sequence[int]): Number of blocks in each stage.
+            - width_factor (Sequence[float]): Width deflator in each stage.
+            - num_conv_branches (Sequence[float]): Number of conv branches
+                 in each stage.
```
Comment (mzr1996) [inline mmcls/models/backbones/mobileone.py:280]: The item in a list should be aligned left.
```suggestion
            - num_conv_branches (Sequence[float]): Number of conv branches
              in each stage.
```

File: tools/convert_models/mobileone_to_mmcls.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                new_k = k
+        new_k = 'backbone.' + new_k
+        new_ckpt[new_k] = new_v
+    return new_ckpt
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description='Convert keys in timm pretrained vit models to '
+        'MMClassification style.')
```
Comment (mzr1996) [inline tools/convert_models/mobileone_to_mmcls.py:44]: Fix the description

File: tools/convert_models/mobileone_to_mmcls.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    parser.add_argument('src', help='src model path or url')
+    # The dst path must be a full path of the new checkpoint.
+    parser.add_argument('dst', help='save path')
+    args = parser.parse_args()
+
+    checkpoint = CheckpointLoader.load_checkpoint(args.src, map_location='cpu')
+
+    if 'state_dict' in checkpoint:
+        # timm checkpoint
```
Comment (mzr1996) [inline tools/convert_models/mobileone_to_mmcls.py:53]: remove the comment

File: tools/convert_models/mobileone_to_mmcls.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+import argparse
+import os.path as osp
+from collections import OrderedDict
+
+import mmcv
+import torch
+from mmcv.runner import CheckpointLoader
+
+
+def convert_twins(args, ckpt):
```
Comment (mzr1996) [inline tools/convert_models/mobileone_to_mmcls.py:11]: convert `twins`?

**Review by mzr1996 — CHANGES_REQUESTED** [review]

**Comment by mzr1996** [comment]
Please add it to the `model-zoo.md` and `models.rst`.

**Comment by Ezra-Yu** [comment]
close it due to the branch dev-a.x have supported it.

**Pull request closed without merging**